internal/unpackinfo/unpackinfo.go

```diff
@@ -30,20 +30,21 @@ type UnpackInfo struct {
 // It will return an error if the header represents an illegal symlink extraction
 // or if the entry type is not supported by go-slug.
 func NewUnpackInfo(dst string, header *tar.Header) (UnpackInfo, error) {
-	// Get rid of absolute paths.
-	path := header.Name
-
-	if path[0] == '/' {
-		path = strings.TrimPrefix(path, "/")
+	// Check for empty destination
+	if len(dst) == 0 {
+		return UnpackInfo{}, errors.New("empty destination is not allowed")
 	}
-	path = filepath.Join(dst, path)
 
-	// Check for paths outside our directory, they are forbidden
-	if len(dst) > 0 && !strings.HasSuffix(dst, "/") {
-		dst += "/"
-	}
+	// Clean the destination path
+	dst = filepath.Clean(dst)
+	path := filepath.Clean(header.Name)
+
+	path = filepath.Join(dst, path)
 	target := filepath.Clean(path)
-	if !strings.HasPrefix(target, dst) {
+
+	// Check for path traversal by ensuring the target is within the destination
+	rel, err := filepath.Rel(dst, target)
+	if err != nil || strings.HasPrefix(rel, "..") || strings.Contains(dst, "..") || strings.Contains(path, "..") {
```

Comment:
Following up on my other comment, this would align us with Nomad's logic, which seems sufficient / simpler?

https://github.com/hashicorp/nomad/blob/b7aabb11be1c5b7616d82f35553cce8b01cc187b/helper/escapingfs/escapes.go#L107-L119

```suggestion
	if err != nil || strings.HasPrefix(rel, "..") {
```

Reply:
Good point, also thanks for reviewing it on Slack 🙌  Since the `dst` is set by us and is not a user provided data I'll go ahead and approve the suggestion and edit the destination related tests